File: lzw15v.cs
// Bradford Arrington 2025
using System;
using System.IO;
using System.Collections.Generic;

partial class Compressor
{
    public string CompressionName = "LZW 15 Bit Variable Rate Encoder";
    public static string Usage = "in-file out-file\n\n";
    private const int Bits = 15;
    private const int MaxCode = (1 << Bits) - 1;
    private const int TableSize = 35023;
    private const int TableBanks = (TableSize >> 8) + 1;
    private const int EndOfStream = 256;
    private const int BumpCode = 257;
    private const int FlushCode = 258;
    private const int FirstCode = 259;
    private const int Unused = -1;

    private readonly Dictionary<int, (int CodeValue, int ParentCode, char Character)>[] dict;
    private readonly char[] decodeStack = new char[TableSize];

    private int nextCode;
    private int currentCodeBits;
    private int nextBumpCode;

    public Compressor()
    {
        dict = new Dictionary<int, (int CodeValue, int ParentCode, char Character)>[TableBanks];
    }

    private  void InitializeStorage()
    {
        for (int i = 0; i < TableBanks; i++)
        {
            dict[i] = new Dictionary<int, (int CodeValue, int ParentCode, char Character)>(256);
        }
    }

    public  void InitializeDictionary()
    {
        for (int i = 0; i < TableSize; i++)
        {
            int bank = i >> 8;
            int index = i & 0xff;

            if (!dict[bank].ContainsKey(index))
            {
                dict[bank][index] = (Unused, -1, '\0');
            }
        }

        nextCode = FirstCode;
        Console.Write('F'); // Represents putc('F', stdout)
        currentCodeBits = 9;
        nextBumpCode = 511;
    }
    private  int FindChildNode(int parentCode, int childCharacter)
    {
        int index = (childCharacter << (Bits - 8)) ^ parentCode;
        int offset = (index == 0) ? 1 : TableSize - index;

        while (true)
        {
            int bank = index >> 8;
            int subIndex = index & 0xff;

            if (!dict[bank].ContainsKey(subIndex) || dict[bank][subIndex].CodeValue == Unused)
            {
                return index;
            }

            var entry = dict[bank][subIndex];
            if (entry.ParentCode == parentCode && entry.Character == (char)childCharacter)
            {
                return index;
            }

            if (index >= offset)
            {
                index -= offset;
            }
            else
            {
                index += TableSize - offset;
            }
        }
    }

    private  uint DecodeString(uint count, uint code)
    {
        while (code > 255)
        {
            int bank = (int)(code >> 8);
            int subIndex = (int)(code & 0xff);

            if (!dict[bank].TryGetValue(subIndex, out var entry))
            {
                throw new Exception("Invalid code during decoding.");
            }

            decodeStack[count++] = entry.Character;
            code = (uint)entry.ParentCode;
        }

        decodeStack[count++] = (char)code;
        return count;
    }
    public void CompressFile(Stream input, BitFile output, int argc, string[] argv)
    {
        int character;
        int stringCode;
        int index;

        // Initialize the dictionary and storage
        InitializeStorage();
        InitializeDictionary();

        // Read the first character
        stringCode = input.ReadByte();
        if (stringCode == -1) // Handle end-of-stream case
        {
            stringCode = EndOfStream;
        }

        // Process each character in the input stream
        while ((character = input.ReadByte()) != -1)
        {
            // Find or add the string in the dictionary
            index = FindChildNode(stringCode, character);

            if (dict[index >> 8].ContainsKey(index & 0xff) &&
                dict[index >> 8][index & 0xff].CodeValue != Unused)
            {
                stringCode = dict[index >> 8][index & 0xff].CodeValue;
            }
            else
            {
                // Add the new string to the dictionary
                int bank = index >> 8;
                int subIndex = index & 0xff;
                dict[bank][subIndex] = (nextCode++, stringCode, (char)character);

                // Output the code for the string
                output.OutputBits((uint)stringCode, currentCodeBits);

                // Update the current string to the current character
                stringCode = character;

                // Handle dictionary size limits
                if (nextCode > MaxCode)
                {
                    // Output the flush code and reinitialize the dictionary
                    output.OutputBits((uint)FlushCode, currentCodeBits);
                    InitializeDictionary();
                }
                else if (nextCode > nextBumpCode)
                {
                    // Output the bump code and increase the code size
                    output.OutputBits((uint)BumpCode, currentCodeBits);
                    currentCodeBits++;
                    nextBumpCode = (nextBumpCode << 1) | 1;
                    Console.Write('B'); // Pacifier character
                }
            }
        }

        // Output the final string and the end-of-stream code
        output.OutputBits((uint)stringCode, currentCodeBits);
        output.OutputBits((uint)EndOfStream, currentCodeBits);

        while (argc-- > 0)
            Console.WriteLine($"Unknown argument: {argv[argv.Length - argc - 1]}");
    }
    public void ExpandFile(BitFile input, Stream output, int argc, string[] argv)
    {
        uint newCode;
        uint oldCode;
        int character;
        uint count;

        InitializeStorage();

        while (argc-- > 0)
            Console.WriteLine($"Unknown argument: {argv[argv.Length - argc - 1]}");

        while (true)
        {
            InitializeDictionary();

            // Read the first code
            oldCode = (uint)input.InputBits(currentCodeBits);
            if (oldCode == EndOfStream)
            {
                return;
            }

            character = (int)oldCode;
            output.WriteByte((byte)oldCode);

            while (true)
            {
                // Read the next code
                newCode = (uint)input.InputBits(currentCodeBits);
                if (newCode == EndOfStream)
                {
                    return;
                }

                if (newCode == FlushCode)
                {
                    break;
                }

                if (newCode == BumpCode)
                {
                    currentCodeBits++;
                    Console.Write('B');
                    continue;
                }

                if (newCode >= nextCode)
                {
                    decodeStack[0] = (char)character;
                    count = DecodeString(1, oldCode);
                }
                else
                {
                    count = DecodeString(0, newCode);
                }

                character = decodeStack[count - 1];

                // Write decoded string to output
                while (count > 0)
                {
                    output.WriteByte((byte)decodeStack[--count]);
                }

                // Add the new entry to the dictionary
                int bank = (int)(nextCode >> 8);
                int subIndex = (int)(nextCode & 0xff);
                dict[bank][subIndex] = (nextCode++, (int)oldCode, (char)character);

                // Update old code
                oldCode = newCode;
            }
        }
    }
}

File: Python/lzw15v.py
# Bradford Arrington 2025
import sys
from io import FileIO, SEEK_SET, SEEK_CUR
from collections import defaultdict
from bitio import CompressorBitio # as CompressorBitio
import time
import tracemalloc
import psutil
import os
import struct
from typing import BinaryIO, List, Tuple, Optional
    
class Compressor_lzw15v:
    COMPRESSION_NAME = "LZW 15 Bit Variable Rate Encoder"
    USAGE = "in-file out-file\n\n"
    BITS  =                     15
    MAX_CODE =                  ( ( 1 << BITS ) - 1 )
    TABLE_SIZE =                35023
    TABLE_BANKS =               ( ( TABLE_SIZE >> 8 ) + 1 )
    END_OF_STREAM =              256
    BUMP_CODE  =                257
    FLUSH_CODE =                258
    FIRST_CODE =                259
    UNUSED  =                   -1

    def __init__(self):
        self.dict = [None] * self.TABLE_BANKS
        self.decode_stack = [0] * self.TABLE_SIZE
        self.next_code = self.FIRST_CODE
        self.current_code_bits = 9
        self.next_bump_code = 511
        self.initialize_storage()

    def initialize_storage(self):
        """Initialize the dictionary storage"""
        for i in range(self.TABLE_BANKS):
            self.dict[i] = [{'code_value': self.UNUSED, 'parent_code': 0, 'character': 0} 
                           for _ in range(256)]

    def initialize_dictionary(self):
        """Initialize/reset the dictionary"""
        for i in range(self.TABLE_SIZE):
            self.dict_lookup(i)['code_value'] = self.UNUSED
        self.next_code = self.FIRST_CODE
        print('F', end='', flush=True)
        self.current_code_bits = 9
        self.next_bump_code = 511

    def dict_lookup(self, index: int) -> dict:
        """Dictionary lookup macro equivalent"""
        bank = index >> 8
        offset = index & 0xFF
        if self.dict[bank] is None:
            raise RuntimeError(f"Dictionary bank {bank} not initialized (index={index})")
        return self.dict[bank][offset]

    def decode_string(self, count: int, code: int) -> int:
        """
        Decode a string from the dictionary and store it in decode_stack
        Returns the count of characters placed in the stack
        """
        while code > 255:
            dict_entry = self.dict_lookup(code)
            self.decode_stack[count] = dict_entry['character']
            count += 1
            code = dict_entry['parent_code']
        self.decode_stack[count] = code
        return count + 1

    def find_child_node(self, parent_code: int, child_character: int) -> int:
        """
        Find the table location for a string/character combination
        Uses XOR hashing with collision handling
        """
        index = (child_character << (self.BITS - 8)) ^ parent_code
        if index == 0:
            offset = 1
        else:
            offset = self.TABLE_SIZE - index
            
        while True:
            dict_entry = self.dict_lookup(index)
            if dict_entry['code_value'] == self.UNUSED:
                return index
            if (dict_entry['parent_code'] == parent_code and 
                dict_entry['character'] == child_character):
                return index
            if index >= offset:
                index -= offset
            else:
                index += self.TABLE_SIZE - offset

    def compress_file(self, input_stream: FileIO, output: 'CompressorBitio.BitFile', argc: int, argv: list):
        character: int = 0
        string_code: int = 0
        index: int = 0

        # Initialize dictionary
        #self.initialize_storage() # handled in __init__
        self.initialize_dictionary()

        char_byte = input_stream.read(1)
        if not char_byte:  # EOF
            string_code = self.END_OF_STREAM
        else:
            string_code = char_byte[0]

        while True:
            char_byte = input_stream.read(1)
            if not char_byte:  # EOF
                break

            character = char_byte[0]

            index = self.find_child_node( string_code, character)
            dict_entry = self.dict_lookup(index)
            
            if dict_entry['code_value'] != self.UNUSED:
                string_code = dict_entry['code_value']
            else:
                dict_entry['code_value'] = self.next_code
                dict_entry['parent_code'] = string_code
                dict_entry['character'] = character

                output.output_bits( string_code, self.current_code_bits )
                string_code = character
                self.next_code += 1

                if self.next_code > self.MAX_CODE:
                    output.output_bits( self.FLUSH_CODE, self.current_code_bits )
                    self.initialize_dictionary()
                elif self.next_code > self.next_bump_code:
                    output.output_bits( self.BUMP_CODE, self.current_code_bits )
                    self.current_code_bits += 1
                    self.next_bump_code <<= 1
                    self.next_bump_code |= 1
                    print( 'B' )

        # Write the last string and end-of-stream marker
        output.output_bits(string_code, self.current_code_bits)
        output.output_bits(self.END_OF_STREAM, self.current_code_bits)
        # flush ??

        while argc > 0:
            argc -= 1
            print(f"Unknown argument: {argv[len(argv) - argc - 1]}")
   
    def expand_file(self, input_bit_file: 'CompressorBitio.BitFile', output_stream: FileIO, argc: int, argv: list[str]):
        """Expand compressed file using LZW algorithm"""
        new_code: int = 0
        old_code: int = 0
        character: int = 0
        count: int = 0
        
        #self.initialize_storage()
        while True:
            self.initialize_dictionary()
            old_code = input_bit_file.input_bits(self.current_code_bits)
            if old_code == self.END_OF_STREAM:
	            break
                
            character = old_code
            output_stream.write(bytes([character]))
            
            while True:
                new_code = input_bit_file.input_bits(self.current_code_bits)
                if new_code == self.END_OF_STREAM:
                    return
                if new_code == self.FLUSH_CODE:
                    break
                if new_code == self.BUMP_CODE:
                    self.current_code_bits += 1
                    print('B', end='', flush=True)
                    continue
                
                if new_code >= self.next_code:
                    self.decode_stack[0] = character
                    count = self.decode_string(1, old_code)
                else:
                    count = self.decode_string(0, new_code)
                
                character = self.decode_stack[count - 1]
                
                # Output the decoded string
                for i in range(count - 1, -1, -1):
                    output_stream.write(bytes([self.decode_stack[i]]))
                
                # Add new entry to dictionary
                if self.next_code < self.TABLE_SIZE:
                    dict_entry = self.dict_lookup(self.next_code)
                    dict_entry['parent_code'] = old_code
                    dict_entry['character'] = character
                    self.next_code += 1
                
                old_code = new_code
